### PR TITLE
Transaction Exempt

### LINKED ISSUE #906: Transaction to modify exempt transactions
# Exempt from transaction fee

The following hardcoded list of transactions are exempt from transaction fee.
- add_miner
- miner_health_check
- add_sharder
- sharder_health_check
- contributeMpk
- sharder_keep
- shareSignsOrShares
- wait

# Change exempt transaction

We will be adding a new transaction that will modify the list of exempt transitions. We need the following changes.

1. Create an enumeration of smart contract endpoints.
2. Add to the genesis settings a list of smart contract transactions that are to be exempt from fees.
3. Add a REST API endpoint to get the list of transactions that are currently exempt from fees.
4. Add a new transaction that will either change a transaction to be exempt from fees or will change a transaction to be subject to fees.

# Genesis settings

We should add to the genesis yaml file the initial list of exempt transactions.
```yaml
transaction:
  payload:
    max_size: 98304 # bytes
  timeout: 30 # seconds
  min_fee: 0
  exempt:
  - add_miner
  - miner_health_check
  - add_sharder
  - sharder_health_check
  - contributeMpk
  - sharder_keep
  - shareSignsOrShares
  - wait
  - add_authoizer
  - blobber_health_check
```

### PR DESCRIPTION
**Pull request by ssardana96**
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feat/exempt_transaction  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/chaincore/transaction/entity.go [hunk 1]
```diff
@@ -94,19 +94,19 @@ type smartContractTransactionData struct {
...
+//var exemptedSCFunctions = map[string]bool{
+//	"add_miner":            true,
+//	"miner_health_check":   true,
+//	"add_sharder":          true,
+//	"sharder_health_check": true,
+//	"contributeMpk":        true,
+//	"sharder_keep":         true,
+//	"shareSignsOrShares":   true,
+//	"wait":                 true,
+//}
```
Comment (Sriep) [inline code/go/0chain.net/chaincore/transaction/entity.go:106]: You can delete this comment now. Moved to 0chain.yaml.

Comment (Sriep) [inline code/go/0chain.net/chaincore/transaction/entity.go]: You can delete this now, information moved to 0chain.yaml.

**Review by Sriep — DISMISSED** [review]
Looks ok for the moment. We might need to modify later to handle duplicate function names.

Integration testing.
1. To start just test global-config and global-update-config. So your not testing the fee-charging bit, just that the setting can be changed and is integrated correctly.
2. Next we need a way of eyeballing exempt fee in action. Maybe try to not give yourself enough tokens to pay for the transactions. And check you fail at the correct time.

**Review by ssardana96 — COMMENTED** [review]
tested and verified

**Comment by ssardana96** [comment]
> Looks ok for the moment. We might need to modify later to handle duplicate function names.
> 
> Integration testing.
> 
> 1. To start just test global-config and global-update-config. So your not testing the fee-charging bit, just that the setting can be changed and is integrated correctly.
> 2. Next we need a way of eyeballing exempt fee in action. Maybe try to not give yourself enough tokens to pay for the transactions. And check you fail at the correct time.

done

**Review by ssardana96 — COMMENTED** [review]
done

**Review by Sriep — APPROVED** [review]

**Pull request merged**